gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/GitConfigMonitor.java

```diff
@@ -0,0 +1,419 @@
+    } catch (IOException e) {
+      throw new RuntimeException("Could not create pull file loader", e);
+    }
+
+    try {
+      this.gitRepo = new GitRepository(repositoryUri, this.repositoryDir);
+    } catch (GitAPIException | IOException e) {
+      throw new RuntimeException("Could not open git repository", e);
+    }
+  }
+
+  /** Start the service. */
+  @Override
+  protected void startUp() throws Exception {
+    log.info("Starting the " + GitConfigMonitor.class.getSimpleName());
+    log.info("Polling git with inteval {} ", this.pollingInterval);
+
+    // Schedule the job config fetch task
+    this.scheduledExecutor.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          processGitConfigChanges();
+        } catch (GitAPIException | IOException e) {
+          log.error("Failed to process git config changes", e);
+          // next run will try again since errors could be intermittent
+        }
+      }
+    }, 0, this.pollingInterval, TimeUnit.SECONDS);
+  }
+
+  /** Stop the service. */
+  @Override
+  protected void shutDown() throws Exception {
+    this.scheduledExecutor.shutdown();
+    this.scheduledExecutor.awaitTermination(TERMINATION_TIMEOUT, TimeUnit.SECONDS);
+  }
+
+  public synchronized void setActive(boolean isActive) {
+    if (this.isActive == isActive) {
+      // No-op if already in correct state
+      return;
+    }
+
+    this.isActive = isActive;
+  }
+
+  /**
+   * Fetch the list of changes since the last refresh of the repository and apply the changes to the {@link FlowCatalog}
+   * @throws GitAPIException
+   * @throws IOException
+   */
+  @VisibleForTesting
+  void processGitConfigChanges() throws GitAPIException, IOException {
+    // if not active or if the flow catalog is not up yet then can't process config changes
+    if (!isActive || !this.flowCatalog.isRunning()) {
+      return;
+    }
+
+    List<DiffEntry> changes = this.gitRepo.getChanges();
+
+    for (DiffEntry change : changes) {
+      switch (change.getChangeType()) {
+        case ADD:
+        case MODIFY:
+          addSpec(change);
+          break;
+        case DELETE:
+          removeSpec(change);
+          break;
+        case RENAME:
+          removeSpec(change);
+          addSpec(change);
+          break;
+        default:
+          throw new RuntimeException("Unsupported change type " + change.getChangeType());
+      }
+    }
+
+    // Done processing changes, so checkpoint
+    this.gitRepo.moveCheckpointAndHashesForward();
+  }
+
+  /**
+   * Add a {@link FlowSpec} for an added, updated, or modified flow config
+   * @param change
+   * @throws IOException
+   */
+  private void addSpec(DiffEntry change) throws IOException {
+    if (checkConfigFilePath(change.getNewPath())) {
+      Path configFilePath = new Path(this.repositoryDir, change.getNewPath());
+      Config flowConfig = loadConfigFileWithFlowNameOverrides(configFilePath);
```

Comment:
Should we handle exception here if the config file is malformed and simply log the error instead of bubbling up?